### PR TITLE
Handle Coalesce expressions in control flow graph

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -203,6 +203,10 @@ Microsoft.CodeAnalysis.Operations.CommonConversion.IsNumeric.get -> bool
 Microsoft.CodeAnalysis.Operations.CommonConversion.IsReference.get -> bool
 Microsoft.CodeAnalysis.Operations.CommonConversion.IsUserDefined.get -> bool
 Microsoft.CodeAnalysis.Operations.CommonConversion.MethodSymbol.get -> Microsoft.CodeAnalysis.IMethodSymbol
+Microsoft.CodeAnalysis.Operations.ConditionalBranchKind
+Microsoft.CodeAnalysis.Operations.ConditionalBranchKind.IfFalse = 1 -> Microsoft.CodeAnalysis.Operations.ConditionalBranchKind
+Microsoft.CodeAnalysis.Operations.ConditionalBranchKind.IfNull = 2 -> Microsoft.CodeAnalysis.Operations.ConditionalBranchKind
```
Comment (gafter) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: >CodeAnalysis [](start = 10, length = 12)

This feels out of place. Conditions should be boolean; we have existing expression nodes that can express a null test. Also it seems strange to have an IfNull but not an IfNotNull.

**Review by gafter — COMMENTED** [review]
`ConditionalBranchKind` doesn't feel right.

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: > This feels out of place. Conditions should be boolean; we have existing expression nodes that can express a null test. Also it seems strange to have an IfNull but not an IfNotNull.

I see nothing wrong to have an easy way to represent a null check in the graph. Of course, we'll discuss that at the design meeting. But for now I am changing Condition to a Value and adding this new branch kind. If any follow-up is needed based on resolution of the meeting, I'll do them in a separate PR

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: For now IfNotNull is not needed, but we can add it for completeness. Again let's have conversation at the design meeting.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -521,19 +526,119 @@ private void VisitConditionalBranch(IOperation condition, ref BasicBlock dest, b
...
         }
 
+        public override IOperation VisitCoalesce(ICoalesceOperation operation, int? captureIdForResult)
+        {
+            SyntaxNode valueSyntax = operation.Value.Syntax;
+            ITypeSymbol valueTypeOpt = operation.Value.Type;
+
+            SpillEvalStack();
+
+            int save_availableCaptureId = _availableCaptureId;
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Perhaps `previousAvailableCaptureId` or `saveAvailableCaptureId`, to match naming convention.

**Review by cston — APPROVED** [review]

**Review by mavasani — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (gafter) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: I cannot attend the design meeting due to recovering from surgery.

We have conditionals in many places in the languages C# and VB, and in none of these is there a primitive "null check" as a condition (as there is in C). Neither is there such a thing in either compiler's lowered form: we instead use a boolean expression which checks the input against null. This seems to be an arbitrarily specific combination of the equality and null primitives, and lowered forms should not be introducing such higher-level operations.

**Review by 333fred — APPROVED** [review]

**Review by gafter — APPROVED** [review]
Looks good to me, modulo the API issue that I recorded in #24436.

**Pull request merged**